evaluator.plandefinition/src/main/java/org/opencds/cqf/cql/evaluator/plandefinition/r4/PlanDefinitionProcessor.java

```diff
@@ -0,0 +1,725 @@
+    return null;
+  }
+
+  protected void ensureConditionExpression(PlanDefinition.PlanDefinitionActionConditionComponent condition) {
+    if (!condition.hasExpression()) {
+      logger.error("Missing condition expression");
+      throw new RuntimeException("Missing condition expression");
+    }
+  }
+
+  protected void ensureDynamicValueExpression(PlanDefinition.PlanDefinitionActionDynamicValueComponent dynamicValue) {
+    if (!dynamicValue.hasExpression()) {
+      logger.error("Missing dynamicValue expression");
+      throw new RuntimeException("Missing dynamicValue expression");
+    }
+  }
+
+  private String ensureStringResult(Object result) {
+    if (!(result instanceof StringType))
+      throw new RuntimeException("Result not instance of String");
+    return ((StringType) result).asStringValue();
+  }
+
+  protected static String getResourceName(CanonicalType canonical) {
+    if (canonical.hasValue()) {
+      String id = canonical.getValue();
+      if (id.contains("/")) {
+        id = id.replace(id.substring(id.lastIndexOf("/")), "");
+        return id.contains("/") ? id.substring(id.lastIndexOf("/") + 1) : id;
+      }
+      return null;
+    }
+
+    throw new RuntimeException("CanonicalType must have a value for resource name extraction");
+  }
+
+  protected Object evaluateConditionOrDynamicValue(String expression, String language, String libraryToBeEvaluated, Session session, List<DataRequirement> dataRequirements) {
+    Object result = null;
+    switch (language) {
+      case "text/cql":
+      case "text/cql.expression":
+      case "text/cql-expression": {
+        if (dataRequirements != null && !dataRequirements.isEmpty()) {
+          Parameters params = resolveInputParameters(dataRequirements);
+          if (session.parameters != null && session.parameters instanceof Parameters) {
+            ((Parameters) session.parameters).getParameter().forEach(parameter -> params.addParameter(parameter));
+          }
+          expressionEvaluator.evaluate(expression, params);
+        } else {
+          expressionEvaluator.evaluate(expression, session.parameters);
+        }
+      } break;
+      case "text/cql-identifier":
+      case "text/cql.identifier":
+      case "text/cql.name":
+      case "text/cql-name": {
+        Set<String> expressions = new HashSet<>();
+        expressions.add(expression);
+        result = libraryProcessor.evaluate(libraryToBeEvaluated, session.patientId, session.parameters,
+            session.contentEndpoint, session.terminologyEndpoint, session.dataEndpoint, session.bundle, expressions);
+      } break;
+      case "text/fhirpath": {
+        Parameters params = resolveInputParameters(dataRequirements);
+        if (params != null) {
+          IBaseParameters resultParams = expressionEvaluator.evaluate(expression, params);
+          IBaseDatatype tempResult = operationParametersParser.getValueChild(((Parameters) resultParams), "return");
```

Comment:
Within this class we can operate on R4 parameters directly. There's a gap in the parameters resource API, but it's probably more easily covered with a helper like so:
```java
static class ParameterUtil {
   ParameterParametersComponent getParameterComponentByName(Parameters params, String name) {
        return params.getParameters().stream().filter(x -> x.getName().equals(name)).firstOrNull();
   }
}
```

Then you can do a static import on that function.

Then the rest of this code is greatly simplified because you have direct access to the value:

```java
IBase result = ppc.hasValue() ? ppc.getValue() : ppc.getResource();
```